### PR TITLE
Multi conductor prod

### PR DESCRIPTION
**Pull request by steveej**

### CODE REVIEW

**Review by neonphog — APPROVED** [review]
I can't tell what's changed due to the formatting - if there's something you care about actually being reviewed, please point it out.

File: multiplexer/src/index.ts [hunk 1]
```diff
@@ -801,153 +849,168 @@ app.get("/fail", async (req: Request, res: Response) => {
...
   }
 
-  globWss = []
-  for (let i=0; i< CONDUCTOR_COUNT; i+=1) {
-    globWss.push(makeWsServer(i))
+  try {
+    const adminWebsocket = await getAdminWebsocket(i);
+    console.log(`Starting app interface on port ${realAppPortForInterface(i)}`);
+    await adminWebsocket.attachAppInterface({
+      port: realAppPortForInterface(i),
```
Comment (steveej) [inline multiplexer/src/index.ts:1009]: @neonphog

File: multiplexer/src/index.ts [hunk 2]
```diff
@@ -70,40 +70,69 @@ const INSTANCE_COUNT = parseInt(
...
 const conductorForRegKey = (regkey: string): number => {
-  return (Buffer.from(regkey)[1] % CONDUCTOR_COUNT);
+  return Buffer.from(regkey)[0] % CONDUCTOR_COUNT;
 };
 
-const lairBin = process.env.LAIR_PATH
+const lairBin = process.env.LAIR_PATH;
 
-const getLairSocket = (conductor:number) => {
+const getLairRootFromEnv = (conductor: number): string => {
```
Comment (steveej) [inline multiplexer/src/index.ts:118]: @neonphog

File: multiplexer/src/index.ts [hunk 3]
```diff
@@ -70,40 +70,69 @@ const INSTANCE_COUNT = parseInt(
...
+    const result = config.match(/driver:\W+type: websocket\W+port: ([0-9]+)/m);
+    if (!result) throw "Unable to find admin port in config";
+    console.log(`Conductor ${i} on admin port ${result[1]}`);
+    array[i] = result[1];
+  }
+
+  return array;
+};
+
+const HC_ADMIN_PORTS: Array<number> = (
```
Comment (steveej) [inline multiplexer/src/index.ts:104]: @neonphog

File: multiplexer/src/index.ts [hunk 4]
```diff
@@ -70,40 +70,69 @@ const INSTANCE_COUNT = parseInt(
...
+const getHcAdminPortsFromEnv = (): Array<string> => {
+  var env = process.env.HC_ADMIN_PORTS;
+  if (env) {
+    return env.split(",");
+  } else {
+    return [];
+  }
+};
+
+const getHcAdminPortsFromDotHc = (): Array<string> => {
```
Comment (steveej) [inline multiplexer/src/index.ts:90]: @neonphog

File: multiplexer/src/index.ts [hunk 5]
```diff
@@ -70,40 +70,69 @@ const INSTANCE_COUNT = parseInt(
...
+      return "";
+    }
+  } catch (e: any) {
+    throw `error creating agent_key ${e.message}`;
+  }
+};
 
-  const lairRoot = process.env.LAIR_WORKING_DIRECTORY || lairWorkdirPathFromDotHC (conductor)
-    // prefer getting the url from lair-keystore directly
+const getLairSocket = (conductor: number) => {
```
Comment (steveej) [inline multiplexer/src/index.ts:131]: @neonphog

**Review by steveej — COMMENTED** [review]
@neonphog that's understandable. i just marked the spots where i introduced semantic changes.

**Review by neonphog — APPROVED** [review]

**Pull request merged**